### ISSUE TITLE
Association end name below line?

### ISSUE THREAD

**Issue by garretwilson** [umlet/umlet#447]
When associating two classes, how can I place the name of an association end (such as `r1`) _below_ the line instead of above it?

See for example the Professor-Book relationship at http://www.uml-diagrams.org/association.html and the names "author" and "textbook":

![](http://www.uml-diagrams.org/association/association.png)

Placing the association end names above the line gets in the way of the name of the association itself (such as "Wrote" in the example. How can I place them below the line?

**Comment by afdia** [comment]
As the `UML Common Elements` palette shows, roles (r1, r2) are shown on top, messages (m1, m2) on bottom of the line.
You can also readjust with r1pos, m1pos, ... but this may be a bit cumbersome

**Comment by garretwilson** [comment]
Why is this closed? Is this not a valid feature request? Are you saying you don't expect to ever address this? Why not?

**Comment by afdia** [comment]
Which feature request? I explained how to put text above and under the line so what is missing?
Placing "Wrote" below the line can be done by putting an empty line above it

**Comment by garretwilson** [comment]
> I explained how to put text above ...

The association names are already above the line by default. That was the reason for filing this issue. I don't want them above the line. I want them below the line; see the example in the first comment.

> ... and under the line ...

Oh, are you saying that I should show the "roles" as "messages"? But isn't that changing the semantics? What do "messages" refer to in the UML model, anyway? (Or are you just using "r1" and "m1" as "above-line" and "below-line" designators, with no semantics?)

**Comment by afdia** [comment]
Ok now I understand that the question is about the semantics and not the visuals.

Typically UMLet doesn't care about semantics; it doesn't have to understand it, because it doesn't generate code or anything based on diagrams. It's just a simple tool to create diagrams in an easy way (mostly but not only for UML).

The relation element offers several customizations which may or may not be useful for certain diagrams, so in fact you can see r1 and m1 as placeholders for "above-line" and "below-line".

Even if we would introduce setting like "roles-below=true", "messages-above=true" most people would probably still just replace the r's with m's because it's the easiest way to change the look of the element.

**Comment by garretwilson** [comment]
Why not determine whether a role is rendered above or below the line based upon whether in the properties it is listed above or below the line designation in the properties?

For example to show the roles above the line:

```
r1=author
writes >
r2=book
lt=-
```

And to show the roles below the line:

```
writes >
lt=-
r1=author
r2=book
```

To me that seems like a simple, logical, and elegant solution. You are maintaining the semantics of the properties, yet you are providing hints for the visual display based upon the order in which you list things.

**Comment by afdia** [comment]
You're right that this would be better than introducing a specific command for it, but using the position also has some problem:
* its not backwards compatible and would break all existing diagrams
* what happens if m1 and r1 are both below lt? Do they overlap? Do we show an error?
* How does the positioning effect alignment for vertical relations?
* The command would behave different than other non-repeatable "global" commands such as fg=red, style=autoresize, ... which do the same thing when they are placed on the first or the last line

Some time ago we discussed if we should change commands line fg=red from a global command to "change the color for all following lines", but we ended up not changing it, because the model of global commands has some nice merits (e.g. you can place them anywhere - e.g. clustering them at the bottom of the properties)

Therefore I'm not sure if including the position in this case is better than just staying with the current way of saying `r1/r2/m1/m2 are generic relation descriptions with a default position which may be adjusted using r1pos/...`

If we plan on releasing a version with breaking changes, we will discuss this change (therefore I will rename the issue and reopen it until we have settled on a solution).

**Comment by garretwilson** [comment]
> using the position also has some problem

It was just an idea; we should think of other ideas as well. My larger point is that forcing users to use a label that _means_ something else just to get a label to show up in the desired place is an even worse approach semantically.

> what happens if m1 and r1 are both below lt? Do they overlap?

Yes, but keep in mind it's no different than if you indicate a role name and an association currently---they will overlap unless you make the horizontal line really long. In fact that's why I opened this issue in the first place.

> ...  saying r1/r2/m1/m2 are generic relation descriptions ...

I understand what you're saying, but that grates against my approach to design. We might as well name them "tl" (top-left), "bl" (bottom-left), etc. Since the whole point of UML is modeling semantics, it would be a shame to turn it into just a line drawing tool.

Then again, maybe it _is_ just a line-drawing tool. Maybe I'm trying to mold it into something it isn't (but could have been).
